add stockpiled button on request sidebar button

dvincent56 opened the pull request:
Revamp the request button in sidebar to add feature store/stop storing resource and send it asap.

If user has no enough resources, button is 'store' / 'stop storing' depending storage policy of the resource.
![store](https://user-images.githubusercontent.com/81165266/114324293-9292f100-9b29-11eb-8e69-e5b4875ec756.png)
![stop_storing](https://user-images.githubusercontent.com/81165266/114324299-945cb480-9b29-11eb-9367-4d41b9c680fe.png)

If user has enough resource, button is 'send request'
![send](https://user-images.githubusercontent.com/81165266/114324301-96267800-9b29-11eb-8b09-386f3997d49a.png)

Asap resources are ok, button switch from 'store' or 'stop storing' to 'send'

If user press 'send', the request is sended to Rome and if resource was storepiled before, it switch to 'stop storing' automatically.

- I haven't test with RESOURCE_DENARII
- I haven't test with RESOURCE_TROOPS

Theoretically, only button 'send request' is available for these resources

crudelios commented:
Some ideas:

- I'd rather have the "current amount/required" in white, except if it's stockpiling, where current amount could be in red.
- It is unclear what the number (`22`) means. An alternative is to draw everything as `13/15, 22 Months` and ellipsise `Months` if it doesn't fit. Also, make its colour white again except when there are 3 or fewer months left, so the user can better see the changes.

I'm slightly color blind and have difficulty distinguishing green text from red text, whence why I prefer white vs red or white vs green.

dvincent56 commented:
I put "11/15" in black to fit with other information in the panel
Title is white, information is black
![sidebar](https://user-images.githubusercontent.com/81165266/114357276-da3d6b00-9b71-11eb-8a58-848af70d8827.png)

We can set '11/15' in red and pass to green when it reach 15/15 like other rating

I agree to display 22 months instead of (22)

crudelios commented:
I still think red should appear only while stockpiling, to at least give a visual cue. Other than that, you're right, it's more visually consistent this way.

dvincent56 commented:
I updated the PR

I added a stockpiled icon feedback instead of red color for text because I prefer keep red color if not enough resources like other rating in panel

I am satisfied with the result
![store_icon_feedback](https://user-images.githubusercontent.com/81165266/114392075-9447cd80-9b98-11eb-9216-0a9bd1cd378d.png)

crudelios commented:
This is not working at all on multiple requests:

![city 2021-04-12 13 42 25](https://user-images.githubusercontent.com/6518876/114396104-133b0700-9b95-11eb-95db-e3ea58006cdb.png)

[Here's a test save for you](http://public.josecadete.net/C3/request.sav)

dvincent56 commented:
This is what I was worried about, do you have a save file with these requests so I can fix?

dvincent56 commented:
This is what I was worried about, thx for the save file. I will fix it.

dvincent56 commented:
@crudelios updated design for more requests.
![more_request](https://user-images.githubusercontent.com/81165266/114463178-625c5880-9be4-11eb-99ea-d66dcc92bafa.png)


For your information. Don't know if it's normal but game crash when focusing this panel (on release branch too)
![crash](https://user-images.githubusercontent.com/81165266/114463218-6e481a80-9be4-11eb-9a43-e6dffeeb8fb4.png)

crudelios commented:
Thanks for the warning, crash is fixed now in fa682b6

crudelios commented:
Can you fix the conflicts so I can check this, please?

crudelios commented on src/translation/english.c:
```diff
@@ -585,10 +585,13 @@ static translation_string all_strings[] = {
+    {TR_SIDEBAR_EXTRA_REQUESTS_SEND, "Send request"},
+    {TR_SIDEBAR_EXTRA_REQUESTS_STOCK, "Store"},
+    {TR_SIDEBAR_EXTRA_REQUESTS_UNSTOCK, "Stop storing"},
+    {TR_SIDEBAR_EXTRA_REQUESTS_MONTHS, "Months"}
```
You don't really need this string as base C3 already has one. Check below.

dvincent56 replied:
I don't know how to use native string, where can I find the key correspondence ?

crudelios commented on src/widget/sidebar/extra.c:
```diff
@@ -385,44 +394,100 @@ static int draw_request_buttons(int y_offset)
         if (data.visible_requests < data.active_requests && i == data.visible_requests - 1) {
             text_draw_centered(translation_for(TR_SIDEBAR_EXTRA_REQUESTS_VIEW_ALL),
-                data.x_offset, y_offset + EXTRA_INFO_VERTICAL_PADDING + 4, data.width, FONT_NORMAL_GREEN, 0);
+                data.x_offset, y_offset + 25, data.width, FONT_NORMAL_GREEN, 0);
```
The "view all" button should be vertically centered and slightly bigger.

dvincent56 replied:
yes you can modify

crudelios commented on src/widget/sidebar/extra.c:
```diff
@@ -385,44 +394,100 @@ static int draw_request_buttons(int y_offset)
+                if (status) {
+                    if (status == CITY_REQUEST_STATUS_NOT_ENOUGH_RESOURCES) {
+                        if (is_stockpiled) {
+                            image_draw(assets_get_image_id(assets_get_group_id("Areldir", "UI_Elements"),
```
I know I suggested a visual feedback when goods were being stockpiled, but I was kinda dumb when I made that suggestion.

If goods are stockpiled, the button already displays "stop stockpiling"! So this small icon serves no real purpose...

dvincent56 replied:
The icon add a feedback to the user. Personally i like this feedback, it's more clear than only text on button.

crudelios replied:
We can keep it, sure!

crudelios commented on src/widget/sidebar/extra.c:
```diff
@@ -385,44 +394,100 @@ static int draw_request_buttons(int y_offset)
+            // request time left
+            width += text_draw_number(r->time, 0, "", width, y_offset+2,
+                                      r->time <= REQUEST_MONTHS_LEFT_FOR_RED_WARNING ? FONT_NORMAL_RED : FONT_NORMAL_GREEN);
+            text_draw_ellipsized(translation_for(TR_SIDEBAR_EXTRA_REQUESTS_MONTHS), width - 2, y_offset+2,
```
You could use `lang_text_draw_ellipsized(8, 4 + (r->time != 1), width - 2, y_offset+2,` instead, which will use the built-in strings `Month` or `Months` depending on the number of months remaining.

dvincent56 replied:
yes I will fix it

crudelios commented on src/widget/sidebar/extra.c:
```diff
@@ -551,8 +616,11 @@ static void draw_extra_info_buttons(void)
-            button_border_draw(data.x_offset + 2, data.request_buttons_y_offset + i * EXTRA_INFO_HEIGHT_REQUESTS_PANEL,
-                data.width - 4, EXTRA_INFO_HEIGHT_REQUESTS_PANEL, i == data.focused_request_button_id - 1);
+
+            int y_offset = data.request_buttons_y_offset + i * EXTRA_INFO_HEIGHT_REQUESTS_PANEL + 28;
```
You'd need to change a bit of code here as well to make the "View all" larger and vertically centered.

dvincent56 replied:
Yes you can modify

crudelios commented on src/widget/sidebar/extra.c:
```diff
@@ -569,8 +637,8 @@ int sidebar_extra_handle_mouse(const mouse *m)
     if ((data.info_to_display & SIDEBAR_EXTRA_DISPLAY_REQUESTS) &&
-        generic_buttons_handle_mouse(m, data.x_offset, data.request_buttons_y_offset,
-        buttons_emperor_requests, data.visible_requests, &data.focused_request_button_id)) {
+            generic_buttons_handle_mouse(m, data.x_offset, data.request_buttons_y_offset,
```
Here too. Instead of having the `y` button position calculated statically, you'd need to change it dynamically, depending on which button was the `View all` button. If you don't want to do it then keep the button as-is and I'll fix it later.

dvincent56 replied:
yes you can modify after

crudelios reviewed (commented):
Some few comments.

Note that the "reposition view all button" is optional, I can do it if you find it too difficult.

Another idea I had to implement after this PR is when a resource is stockpiled and the player clicks "send goods", the popup window, besides asking for confirmation to send, could have a checkbox (disabled by default) saying `Keep stockpiling resource`.

dvincent56 commented:
Confirmation popup is a good idea but I need more time to add it. Can we do it later?
